include/tlist.h

```diff
@@ -35,50 +36,278 @@ typedef void *timer_handle;
+	timerlist_heap_entry_set(timerlist, entry_pos, replacement_entry);
+
+	/*
+	 * And "remove" last element (= entry)
+	 */
+	timerlist->size--;
+
+	/*
+	 * Up (or down) heapify based on replacement item size
+	 */
+	cmp_entries = timerlist_entry_cmp(replacement_entry, entry);
+
+	if (cmp_entries < 0) {
+		timerlist_heap_sift_up(timerlist, entry_pos);
+	} else if (cmp_entries > 0) {
+		timerlist_heap_sift_down(timerlist, entry_pos);
+	}
+}
+
+/*
+ * Check if heap is valid.
+ * - Shape property is always fullfiled because of storage in array
+ * - Check heap property
+ */
+static inline int
+timerlist_debug_is_valid_heap(struct timerlist *timerlist)
+{
+	size_t i;
+	size_t left_pos, right_pos;
+	struct timerlist_timer *left_entry;
+	struct timerlist_timer *right_entry;
+	struct timerlist_timer *cur_entry;
+
+	for (i = 0; i < timerlist->size; i++) {
+		cur_entry = timerlist_heap_entry_get(timerlist, i);
+
+		left_pos = timerlist_heap_index_left(i);
+		right_pos = timerlist_heap_index_right(i);
+
+		if (left_pos < timerlist->size &&
+		    (left_entry = timerlist_heap_entry_get(timerlist, left_pos),
+		    timerlist_entry_cmp(left_entry, cur_entry) < 0)) {
+			return (0);
+		}
+
+		if (right_pos < timerlist->size &&
+		    (right_entry = timerlist_heap_entry_get(timerlist, right_pos),
+		    timerlist_entry_cmp(right_entry, cur_entry) < 0)) {
+			return (0);
+		}
+	}
+
+	return (1);
+}
+
+/*
+ * Main functions implementation
+ */
 static inline void timerlist_init(struct timerlist *timerlist)
 {
-	qb_list_init(&timerlist->timer_head);
+
+	memset(timerlist, 0, sizeof(*timerlist));
+
+	timerlist->heap_entries = NULL;
 	pthread_mutex_init(&timerlist->list_mutex, NULL);
 	timerlist_hertz = qb_util_nano_monotonic_hz();
 }
 
+static inline void timerlist_destroy(struct timerlist *timerlist)
+{
+	size_t zi;
+
+	pthread_mutex_destroy(&timerlist->list_mutex);
+
+	for (zi = 0; zi < timerlist->size; zi++) {
+		free(timerlist->heap_entries[zi]);
+	}
+	free(timerlist->heap_entries);
+}
+
 static inline int32_t timerlist_add(struct timerlist *timerlist,
 				 struct timerlist_timer *timer)
 {
-	struct qb_list_head *timer_list = 0;
-	struct timerlist_timer *timer_from_list;
-	int32_t found = QB_FALSE;
+	size_t new_size;
+	struct timerlist_timer **new_heap_entries;
 
 	if (pthread_mutex_lock(&timerlist->list_mutex)) {
 		return -errno;
 	}
-	qb_list_for_each(timer_list, &timerlist->timer_head) {
 
-		timer_from_list = qb_list_entry(timer_list,
-						struct timerlist_timer, list);
+	/*
+	 * Check that heap array is large enough
+	 */
+	if (timerlist->size + 1 > timerlist->allocated) {
+		new_size = (timerlist->allocated + 1) * 2;
 
-		if (timer_from_list->expire_time > timer->expire_time) {
-			qb_list_add_tail(&timer->list, timer_list);
-			found = QB_TRUE;
-			break;	/* for timer iteration */
+		new_heap_entries = realloc(timerlist->heap_entries,
+		    new_size * sizeof(timerlist->heap_entries[0]));
+		if (new_heap_entries == NULL) {
+			return -errno;
```

Comment:
`pthread_mutex_unlock` missing so that (currently) `qb_loop_timer_add` caller won't accidentally deadlock?

Reply:
Yup, good catch, thanks!